### DISCUSSION TITLE
How to obtain an intermediate layer output?

### DISCUSSION THREAD

**Discussion by prajwalkr** [snakers4/silero-models#53] [Q&A] [answered]
How do we obtain the output of an intermediate layer of the pre-trained model? For example, the output at the end of the convolution encoder, or the output just after the transformer encoder layers.

**Comment by snakers4** [comment]
We do not provide this feature out of the box
But you can just modify the model code

**Comment by prajwalkr** [comment]
I am unable to find the code for the model. Sorry, I am new to torch hub and torchscript

**Comment by snakers4** [comment]
It is public within the jit containers

**Comment by prajwalkr** [comment]
I am sorry, as I said, I do not know about JIT containers...can you paste any link here, please? Thank you so much for the quick replies!

**Comment by snakers4** [comment]
> can you paste any link here, please? 

This one for example - https://github.com/snakers4/silero-models/blob/master/models.yml#L9
This is what is loaded under the hood

**Comment by snakers4** [comment]
Adding a model with skip connections may be a good idea for a v3
But maintaining `_q` / `onnx` flavors of this would be a chore

**Comment by prajwalkr** [comment]
> This is what is loaded under the hood

Loading the model using `torch.jit.load` and accessing `.code` gives:

`def forward(self,\n    x: Tensor,\n    calibrate: bool=False) -> Tensor:\n  _0 = self.n_fft\n  _1 = self.hop_length\n  _2 = self.win_length\n  _3 = torch.hann_window(self.n_fft, dtype=ops.prim.dtype(x), layout=None, device=ops.prim.device(x), pin_memory=None)\n  x0 = __torch__.torch.functional.stft(x, _0, _1, _2, _3, True, "reflect", False, True, )\n  _4 = torch.slice(x0, 0, 0, 9223372036854775807, 1)\n  _5 = torch.slice(_4, 1, 0, 9223372036854775807, 1)\n  _6 = torch.slice(_5, 2, 0, 9223372036854775807, 1)\n  x_real = torch.select(_6, 3, 0)\n  _7 = torch.slice(x0, 0, 0, 9223372036854775807, 1)\n  _8 = torch.slice(_7, 1, 0, 9223372036854775807, 1)\n  _9 = torch.slice(_8, 2, 0, 9223372036854775807, 1)\n  x_imag = torch.select(_9, 3, 1)\n  _10 = torch.add(torch.pow(x_real, 2), torch.pow(x_imag, 2), alpha=1)\n  x1 = torch.sqrt(_10)\n  x2 = (self.audio_normalize).forward(x1, )\n  x3 = (self.quant).forward(x2, )\n  x4 = (self.encoder).forward(x3, )\n  x5 = (self.dequant).forward(x4, )\n  _11 = self.decoder\n  _12 = torch.contiguous(torch.permute(x5, [2, 0, 1]), memory_format=0)\n  _13 = torch.permute((_11).forward(_12, None, None, ), [1, 2, 0])\n  x6 = torch.contiguous(_13, memory_format=0)\n  x7 = (self.fc).forward(x6, )\n  x8 = torch.contiguous(torch.transpose(x7, 1, 2), memory_format=0)\n  return (self.softmax).forward(x8, )`


How can I get the "non-jit" style code?

**Comment by snakers4** [comment, marked as answer]
> How can I get the "non-jit" style code?

We have not published the original code of our models for a number of reasons
More human readable code can be found just inside of jit containers (they are just pickled zips or vice versa)

**Comment by prajwalkr** [comment]
Okay. So, how can I modify the above non-human readable code to return at a specific point in the forward pass? I just want to return the output before the FC layer.

**Comment by snakers4** [comment]
these models were not really meant to be modified
if you would like to modify them anyway - please look inside of the containers manually (they should look like folders)

**Comment by Sindhu-Hegde** [comment]
Can you please let me know how to get the previous layer's output, just before FC? I want the output from the decoder, i.e, when I call model(input), I want the model to return the decoder's output instead of softmax probabilities.
